linkerd/app/outbound/src/opaq/concrete.rs

```diff
@@ -23,18 +23,26 @@ use linkerd_app_core::{
     tls,
     transport::{self, addrs::*},
     transport_header::SessionProtocol,
-    Error, Infallible, NameAddr,
+    Addr, Error, Infallible, NameAddr,
 };
 use std::{fmt::Debug, net::SocketAddr, sync::Arc};
 use tracing::info_span;
 
 /// Parameter configuring dispatcher behavior.
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
 pub enum Dispatch {
-    Balance(NameAddr, NameAddr, balance::EwmaConfig),
+    Balance(Addr, NameAddr, balance::EwmaConfig),
```

Comment:
I'm curious about this change and its relation to LogicalAddr. Generally speaking, we've used the `profiles::LogicalAddr` type to indicate that a profile has given us back a named address for the service we're talking to; and we furthermore have some guarantees that balancers are always using named addresses.

This change uses a new LogicalAddr to associate with _all_ traffic. The only real use for this I see is that we may use this as part of the endpoint metrics labels, which ignores the non-named variant. Given all of this, I'm curious to what extent we actually want a new LogicalAddr type--it seems like having a `Param<Option<profiles::LogicalAddr>>` is actually what we want -- we only want to set this label value with the legacy logical address returned from profile updates. For any new metrics, we'll want ParentRef and BackendRef labels.

Reply:
@olix0r So based on this comment, I went and looked over all his again. I have put a PR with some proposed changes along what you have suggested: https://github.com/linkerd/linkerd2-proxy/pull/3309

Namely: 

- removed `opaq::LogicalAddr`, we only use the `Addr` type now in routes, etc
- added the ability for the target to give you an `Option<profiles::LogicalAddr> `that is Some only if we are using profiles
- this should allow us to preserve the semantics of the balancer metrics with respect to the concrete/logical labels that are different if we are using traffic splits.

I even think that the addr types on the `opaq::Routes` type should not be an `Addr` but some variation of a `SocketAddr`. After all, could it ever be a `NamedAddr` in there?